### DISCUSSION TITLE
need help regarding localStorage in next.js client component

### DISCUSSION THREAD

**Discussion by jaimeenkoffeekodes** [vercel/next.js#62709] [Help]
### Summary

when I try to use localStorage.getItem("db_name")  in client component of next.js, it is giving me error that ' localStorage is not defined' , does it mean that next provide any build localStorage function or anything regarding handling it ?

**Comment by icyJoseph** [comment]
Hi,

Client Components, are SSR'd:

- https://github.com/reactwg/server-components/discussions/4

> Many tech-influences created confusion with regards to client components

For various reasons, but in your case, the important thing is that you probably have something like:

```js
const [state, setState] = useState(localStorage.getItem("db_name"));
// or
const [state, setState] = useState(() => localStorage.getItem("db_name"));
```

Both would produce errors when running on a server side environment, such as Node.js, because there's no localStorage there.

The right way to do this, is to accept that localStorage is an external API, that you must synchronize with. 

You can do this, by either: 

- Making the component that uses this hook, run client side only
   - A quick and effective way: https://demystifying-rsc.vercel.app/client-components/no-ssr/ 
   - Less preferred way: https://demystifying-rsc.vercel.app/client-components/next-dynamic/
- Synchronizing with the localStorage in a use-effect or use-layout-effect hook
   - Something like this https://github.com/vercel/next.js/discussions/49131#discussioncomment-6365650

**Comment by jaimeenkoffeekodes** [comment]
thanks
